File: ChessLab/Laboratory/LabExport.swift
import Foundation

/// Export d'une série Laboratoire : PGN (toutes les parties concaténées) et
/// CSV (une ligne par partie). Pur et testable.
enum LabExport {
    /// PGN complet : chaque partie reçoit des en-têtes synthétiques
    /// (`[Event]`, `[Round]`, `[White]`, `[Black]`, `[Result]`,
    /// `[Termination]`) et son movetext est clos par le résultat — sans
    /// quoi une partie adjugée (abandon / nulle d'accord) sortait sans
    /// résultat ni joueurs, inexploitable dans un autre outil. Les tags de
    /// position de départ (`[SetUp]`/`[FEN]`) émis par ChessKit pour une
    /// série à FEN personnalisé sont préservés.
    static func pgn(_ games: [LabCompletedGame], settings: LabGameSettings) -> String {
        let aElo = Int(settings.sideAEloSlider)
        let bElo = Int(settings.sideBEloSlider)
        return games.map { game in
            let white = game.aWasWhite ? "A (Elo \(aElo))" : "B (Elo \(bElo))"
            let black = game.aWasWhite ? "B (Elo \(bElo))" : "A (Elo \(aElo))"

            let lines = game.pgn.trimmingCharacters(in: .whitespacesAndNewlines)
                .split(separator: "\n", omittingEmptySubsequences: false)
            let setupTags = lines.filter { $0.hasPrefix("[SetUp") || $0.hasPrefix("[FEN") }
            var movetext = lines.filter { !$0.hasPrefix("[") }
                .joined(separator: "\n")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !movetext.hasSuffix(game.pgnResult) {
                movetext += (movetext.isEmpty ? "" : " ") + game.pgnResult
            }

            var header = """
            [Event "ChessLab Lab"]
            [Round "\(game.index + 1)"]
            [White "\(white)"]
            [Black "\(black)"]
            [Result "\(game.pgnResult)"]
            [Termination "\(game.reasonLabel)"]
            """
            if !setupTags.isEmpty {
                header += "\n" + setupTags.joined(separator: "\n")
            }
            return header + "\n\n" + movetext
        }.joined(separator: "\n\n")
    }

    static func csv(_ games: [LabCompletedGame]) -> String {
        var rows = ["partie,camp_A,resultat,score_A,demi_coups,fin"]
        for game in games {
            let aColor = game.aWasWhite ? "Blanc" : "Noir"
            let scoreA: String
            switch game.labResult {
            case .winA: scoreA = "1"
            case .draw: scoreA = "0.5"
            case .winB: scoreA = "0"
            }
            rows.append("\(game.index + 1),\(aColor),\(game.pgnResult),\(scoreA),\(game.plyCount),\(game.reasonLabel)")
        }
        return rows.joined(separator: "\n")
    }
}

File: android/app/src/main/java/com/chesslab/lab/LabExport.kt
package com.chesslab.lab

/**
 * L'export d'une série : un PGN (toutes les parties bout à bout) et un CSV
 * (une ligne par partie). Pendant de `LabExport.swift`, pur et testable.
 *
 * Chaque partie reçoit des en-têtes SYNTHÉTIQUES et son movetext est clos par
 * le résultat : sans cela, une partie sortait sans joueurs ni score, donc
 * inexploitable ailleurs. Les tags de position de départ (`SetUp`/`FEN`) émis
 * pour une série à FEN personnalisé sont conservés — sans eux, l'autre outil
 * rejouerait les coups depuis la position standard.
 */
object LabExport {

    fun pgn(games: List<LabCompletedGame>, nameA: String, nameB: String): String =
        games.joinToString("\n\n") { game ->
            val white = if (game.aWasWhite) nameA else nameB
            val black = if (game.aWasWhite) nameB else nameA

            val lines = game.pgn.trim().split("\n")
            val setupTags = lines.filter { it.startsWith("[SetUp") || it.startsWith("[FEN") }
            var movetext = lines.filterNot { it.startsWith("[") }.joinToString("\n").trim()
            if (!movetext.endsWith(game.pgnResult)) {
                movetext += (if (movetext.isEmpty()) "" else " ") + game.pgnResult
            }

            val header = buildString {
                append("[Event \"ChessLab Lab\"]\n")
                append("[Round \"${game.index + 1}\"]\n")
                append("[White \"$white\"]\n")
                append("[Black \"$black\"]\n")
                append("[Result \"${game.pgnResult}\"]\n")
                append("[Termination \"${game.reasonLabel}\"]")
                if (setupTags.isNotEmpty()) append("\n" + setupTags.joinToString("\n"))
            }
            "$header\n\n$movetext"
        }

    fun csv(games: List<LabCompletedGame>): String {
        val rows = ArrayList<String>(games.size + 1)
        rows += "partie,camp_A,resultat,score_A,demi_coups,fin"
        for (game in games) {
            val aColor = if (game.aWasWhite) "Blanc" else "Noir"
            val scoreA = when (game.labResult) {
                LabGameResult.winA -> "1"
                LabGameResult.draw -> "0.5"
                LabGameResult.winB -> "0"
            }
            rows += "${game.index + 1},$aColor,${game.pgnResult},$scoreA,${game.plyCount},${game.reasonLabel}"
        }
        return rows.joinToString("\n")
    }
}
